test/unit-tests/type/matrix/DenseMatrix.test.js

```diff
@@ -158,6 +158,34 @@ describe('DenseMatrix', function () {
     it('should throw an error when called with invalid datatype', function () {
       assert.throws(function () { console.log(new DenseMatrix([], 1)) })
     })
+
+    it('should not mutate the input data when creating a Matrix (1)', function () {
+      const data = [[1, 2]]
+      Object.freeze(data)
+
+      const matrix = new DenseMatrix(data) // should not throw "TypeError: Cannot assign to read only property '0' of object '[object Array]'"
+      assert.deepStrictEqual(matrix.valueOf(), [[1, 2]])
+      assert.notStrictEqual(matrix.valueOf(), data)
+    })
+
+    it('should not mutate the input data when creating a Matrix (2)', function () {
+      const nestedMatrix = new DenseMatrix([1, 2])
+      const data = [nestedMatrix]
+
+      const matrix = new DenseMatrix(data)
+      assert.deepStrictEqual(matrix._data, [[1, 2]])
+      assert.deepStrictEqual(data, [nestedMatrix]) // should not have replaced the nestedMatrix in data itself
+    })
+
+    it('should not mutate the input data operating on a Matrix', function () {
+      const data = [[1, 2]]
+
+      const matrix = new DenseMatrix(data)
+      matrix.set([0, 1], 42)
+
+      assert.deepStrictEqual(matrix, new DenseMatrix([[1, 42]]))
+      assert.deepStrictEqual(data, [[1, 2]])
+    })
   })
 
```

Comment:
I'm not sure. It would help guard immutability, but it would be a lot of work. This very PR is a fix inside `DenseMatrix` itself though it was discovered by using `dotDivide`. I do feel like unit testing `DenseMatrix` for this fix is enough, no need for additional tests in `dotDivide` or other functions. When functions use the public API of for example DenseMatrix, this API should be "safe" and there should be no need to test for immutability "upstream". What do you think?

Reply:
Yeah I was mostly thinking of just picking one (or at most two) basic operation(s) (like multiplication) and adding a test that performing that operation doesn't mutate either argument, just as suspenders to the belt of this underlying test. It could be helpful if for example some time in the future DenseMatrix is no longer the default Matrix implementation... up to you.

Reply:
👍 that can definitely do no harm. I've added a unit test for this in `multiply` (it fails with the old version of the `preprocess` function). See commit 1d3970702a4d0599a481523f71a93d4826a4877f.

Reply:
Looks good to me, resolving this conversation.